ibis/expr/operations/udfs.py

```diff
@@ -0,0 +1,109 @@
+from __future__ import annotations
+
+import abc
+import functools
+import inspect
+import typing
+from typing import Any
+
+from public import public
+
+import ibis.expr.datatypes as dt
+import ibis.expr.operations as ops
+import ibis.expr.rules as rlz
+from ibis.common.annotations import attribute
+
+
+@public
+class ScalarUDFNode(ops.Value):
+    pass
+
+
+@public
+class VectorizedScalarUDFNode(ScalarUDFNode):
+    pass
+
+
+class UDF(abc.ABC):
+    def _wrap(self, wrapper, fn=None, *args: Any, **kwargs: Any):
+        if fn is None:
+
+            def decorator(fn):
+                return functools.update_wrapper(wrapper(fn, *args, **kwargs), fn)
+
+            return decorator
+        else:
+            return functools.update_wrapper(wrapper(fn, *args, **kwargs), fn)
+
+    def __call__(self, fn=None, *args: Any, **kwargs: Any) -> Any:
+        return self._wrap(self.make_wrapper, fn, *args, **kwargs)
+
+    def vectorized(self, fn=None, *args: Any, **kwargs: Any) -> Any:
+        return self._wrap(self.make_vectorized_wrapper, fn, *args, **kwargs)
+
+    @abc.abstractmethod
+    def make_vectorized_wrapper(self, fn, *args, **kwargs):
+        ...
+
+    @abc.abstractmethod
+    def make_wrapper(self, fn, *args, **kwargs):
+        ...
+
+
+class ScalarUDF(UDF):
```

Comment:
This seems to be analogous with `UserDefinedFunction` defined in `ibis/expr/udf/vectorized.py`, can we unify them?

Reply:
It's similar, but I think this whole thing needs to be rebuilt. It's pretty scattered across a bunch the backends, as opposed to having an interface that each backend can implement separately.